File: src/strategy/strategy.cs
// Strategy pattern

// To run this code, .NET 5 or above is required.

// Feature interface
public interface IFly
{
    public void Fly();
}

// Feature classes
public class FlyWithWings : IFly
{
    public void Fly()
    {
        Console.WriteLine("I'm flying!");
    }
}

public class FlyNoWay : IFly
{
    public void Fly()
    {
        Console.WriteLine("I cannot fly.");
    }
}

public class FlyRocketPowered : IFly
{
    public void Fly()
    {
        Console.WriteLine("I'm flying with rocket power!");
    }
}

// Feature interface
public interface IQuack
{
    public void Quack();
}

// Feature classes
public class NormalQuack : IQuack
{
    public void Quack()
    {
        Console.WriteLine("Quack!");
    }
}

public class Squeak : IQuack
{
    public void Quack()
    {
        Console.WriteLine("Squeak!");
    }
}

public class MuteQueak : IQuack
{
    public void Quack()
    {
        Console.WriteLine("Quiet!");
    }
}

// Main component abstract class
public abstract class Duck
{
    public IQuack QuackBehavior { get; set; }
    public IFly FlyBehavior { get; set; }

    public abstract void Display();
    
    public void PerformFly()
    {
        FlyBehavior.Fly();
    }

    public void PerformQuack()
    {
        QuackBehavior.Quack();
    }

    public void Swim()
    {
        Console.WriteLine("All ducks can float on water. Even the fake one does.");
    }
}

// Main component classes
public class MallardDuck : Duck
{
    public MallardDuck()
    {
        QuackBehavior = new NormalQuack();
        FlyBehavior = new FlyWithWings();
    }

    public override void Display()
    {
        Console.WriteLine("I'm a mallard duck.");
    }
}

public class ModelDuck : Duck
{
    public ModelDuck()
    {
        FlyBehavior = new FlyNoWay();
        QuackBehavior = new NormalQuack();
    }

    public override void Display()
    {
        Console.WriteLine("I'm a mock duck.");
    }
}


public static class Program
{
    public static void Main(string[] args)
    {
        Duck mallard = new MallardDuck();
        mallard.PerformQuack();
        mallard.PerformFly();


        Duck model = new ModelDuck();
        model.PerformFly();
        model.FlyBehavior = new FlyRocketPowered();
        model.PerformFly();
    }
}

File: src/strategy/strategy.java
// Strategy pattern

// To run this code, Java 17 or above is required.

// Feature interface
interface FlyBehavior {
    public void fly();
}

// Feature classes
class FlyWithWings implements FlyBehavior {
    public void fly() {
        System.out.println("I can fly!");
    }
}

class FlyNoWay implements FlyBehavior {
    public void fly() {
        System.out.println("I cannot fly.");
    }
}

class FlyRocketPowered implements FlyBehavior {
    public void fly() {
        System.out.println("I'm flying with rocket power!");
    }
}

// Feature interface
interface QuackBehavior {
    public void quack();
}

// Feature classes
class NormalQuack implements QuackBehavior {
    public void quack() {
        System.out.println("Quack!");
    }
}

class Squeak implements QuackBehavior {
    public void quack() {
        System.out.println("Squeak!");
    }
}

class MuteQuack implements QuackBehavior {
    public void quack() {
        System.out.println("Quiet!");
    }
}

// Main component abstract class
abstract class Duck {
    FlyBehavior flyBehavior;
    QuackBehavior quackBehavior;

    public void performFly() {
        flyBehavior.fly();
    }

    public void performQuack() {
        quackBehavior.quack();
    }

    public void swim() {
        System.out.println("All ducks can float on water. Even the fake one does.");
    }

    public void setFlyBehavior(FlyBehavior flyBehavior) {
        this.flyBehavior = flyBehavior;
    }

    public void setQuackBehavior(QuackBehavior quackBehavior) {
        this.quackBehavior = quackBehavior;
    }

    public abstract void display();
}

// Main component classes
class MallardDuck extends Duck {
    public MallardDuck() {
        flyBehavior = new FlyWithWings();
        quackBehavior = new NormalQuack();
    }

    @Override
    public void display() {
        System.out.println("I'm a mallard duck");
    }
}

class ModelDuck extends Duck {
    public ModelDuck() {
        flyBehavior = new FlyNoWay();
        quackBehavior = new NormalQuack();
    }

    @Override
    public void display() {
        System.out.println("I'm a model duck.");
    }
}

public class Main {
    public static void main(String[] args) {
        Duck mallard = new MallardDuck();
        mallard.performQuack();
        mallard.performFly();

        Duck model = new ModelDuck();
        model.performFly();
        model.setFlyBehavior(new FlyRocketPowered());
        model.performFly();
    }
}
